flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecSink.scala

```diff
@@ -81,13 +82,26 @@ class BatchExecSink[T](
   override protected def translateToPlanInternal(
       planner: BatchPlanner): Transformation[Any] = {
     val resultTransformation = sink match {
-      case _: RetractStreamTableSink[T] | _: UpsertStreamTableSink[T] =>
-        throw new TableException("RetractStreamTableSink and UpsertStreamTableSink is not" +
-          " supported in Batch environment.")
-
       case streamTableSink: StreamTableSink[T] =>
-        // we can insert the bounded DataStream into a StreamTableSink
-        val transformation = translateToTransformation(withChangeFlag = false, planner)
+        val transformation = streamTableSink match {
+          case _: RetractStreamTableSink[T] =>
+            translateToTransformation(withChangeFlag = true, planner)
+
+          case upsertSink: UpsertStreamTableSink[T] =>
+            UpdatingPlanChecker.getUniqueKeyForUpsertSink(this, planner, upsertSink) match {
+              case Some(keys) =>
+                upsertSink.setIsAppendOnly(false)
+                upsertSink.setKeyFields(keys)
+              case None =>
+                upsertSink.setIsAppendOnly(true)
+                upsertSink.setKeyFields(null)
```

Comment:
I think this should be always in `isAppendOnly` and may have a key fields. Some connector can have an optimizaton based on this, e.g. MySQL can use INSERT INTO rather than INSERT .. ON DUPLICATE KEY.

Reply:
`setIsAppendOnly(false)` should be more safe.
- User maybe use same MySQL table to multiple batch inserts. Update should be safe. 
- In future, maybe primary keys are come from DDL, batch can not ensure that there is no duplicate key.